File: 1-1-is-unique.py
# 1.1 Is Unique: Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?

# Use a set to keep track of which characters we've seen. O(1) runtime. O(N) space complexity.

def is_unique(string):
    seen = set()
    for char in string:
        if char in seen:
            return False
        else:
            seen.add(char)
    return True

print(is_unique('hello')) # False
print(is_unique('world')) # True

File: 1-1-is-unique.js
// 1.1 Is Unique: Implement an algorithm to determine if a string has all unique characters. What if you cannot use additional data structures?

function is_unique(string)
{
    let length = string.length;
    let seen = new Set();
    for (i = 0; i < length; i++)
    {
        let char = string[i];
        if (seen.has(char)) {
            return false;
        } else {
            seen.add(char);
        }        
    }
    return true;
}

console.log(is_unique('halo')); // true
console.log(is_unique('hello')); // false